src/swagger/specGenerator3.ts

```diff
@@ -316,4 +316,12 @@ export class SpecGenerator3 extends SpecGenerator {
   protected getSwaggerTypeForReferenceType(referenceType: Tsoa.ReferenceType): Swagger.BaseSchema {
     return { $ref: `#/components/schemas/${referenceType.refName}` };
   }
+
+  protected getSwaggerTypeForUnionType(type: Tsoa.UnionType) {
+    return { anyOf: type.types.map(x => this.getSwaggerType(x))};
```

Comment:
Why did your last commit change it from oneOf to anyOf? From my understanding of the [swagger documentation](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#anyof) anyOf allows for an object to satisfy parts of either. Thats not how a TypeScript union works. You must satisfy all requirements of one of the types in the union. So this needs to be changed back to oneOf.

Reply:
```typescript
let a: {value1: number} | {value2: number} = {value1: 1, value2: 2}
```

Is valid.

Reply:
It's valid because you're the way you've written it is telling TypeScript that you're asserting that it is a union. I believe you might have uncovered a bug in TypeScript with this example:
```ts
let a: {value1: number} | {value2: number} = {value1: 1, value2: 2}
```

But if you write it out step by step then TypeScript catches it and won't even let you set value2.

See below or in [this playground](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgJIBkIzAFQJ4AOKA3gFDIXIBucANgK4QCMAXMiPQLYBG0ANKQC+pUJFiIUqAErAA5gAtchEuUo0GEAExsOPaENKkwy5AFEAHnE4FaEAKohgAexDIAvGkzZ8RZAB80GQUlIgBuQwQXAGcwZCgIKPpaMDYLKxt7Rxd3ZDJKajpGVmQAZgF89UY2TWQAelrkHxQAcmICjWLdXihQ9qr2Lm7ewWbkYCj2J1i4KKi5EDhuW2QwJxWTZrTrWwdnEGaAOlV8k5P6gHluACsIBFjaYDE6ZE44PGQXWneoogRgGHeAGsQE4AO6uAhQJxEKBgYAJPjIOAgAAmyGalQgoxRTgSk1iEHM41ioHWvk2lm2mT2hyEQA) I made:
```ts
interface ILeftType {
    value1: number,
}
interface IRightType {
    value2: number
}

type ExampleUnion = ILeftType | IRightType;

const result: ExampleUnion = {
    value1: 3,
    value: 2 // Type '{ value1: number; value: number; }' is not assignable to type 'ExampleUnion'.
                //Object literal may only specify known properties, and 'value' does not exist in type 'ExampleUnion'.
}
```

So yes we have to keep it as `oneOf` and the validation will also have to enforce that the input object can only be all of one of the two objects but not both.

Reply:
Reverted in f9ce396

Reply:
FYI: It actually seems like this was a bug that's fixed as of [3.5](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#improved-excess-property-checks-in-union-types):